Source/CombatExtended/Harmony/Harmony_StunHandler.cs

```diff
@@ -12,86 +13,86 @@ namespace CombatExtended.HarmonyCE
+        private const float FLAT_ADAPTATION_TIME_REDUCTION_RATIO = ADAPTATION_TIME_RATIO / 8; // Scaling of adaptation reduction
+        private const float REDUCTION_MULTIPLIER_WHILE_STUNNED = 0.8f;
+        private const float BUILDING_DEFAULT_BODYSIZE = 2f;
 
-		    float stunResistChance = ((float) ___EMPAdaptedTicksLeft / (float) newStunAdaptedTicks) * 15;
-		    void reStun()
-		    {
-			if (___stunTicksLeft > 0 && newStunTicks > ___stunTicksLeft)
-			{
-			    ___stunTicksLeft = newStunTicks;
-			}
-			else
-			{
-			    __instance.StunFor_NewTmp(newStunTicks, dinfo.Instigator, true, true);
-			}
-			
-		    }
-		    
-		    if (UnityEngine.Random.value > stunResistChance)
-		    {
-			___EMPAdaptedTicksLeft += Mathf.RoundToInt(dinfo.Amount * 45 * bodySize);
-			reStun();
-		    }
-		    else
-		    {
-			MoteMaker.ThrowText(new Vector3((float)__instance.parent.Position.x + 1f, (float)__instance.parent.Position.y, (float)__instance.parent.Position.z + 1f), __instance.parent.Map, "Adapted".Translate(), Color.white, -1f);
-			int adaptationReduction = Mathf.RoundToInt(Mathf.Sqrt(dinfo.Amount * 45));
-			
-			if (adaptationReduction < ___EMPAdaptedTicksLeft) {
-			    ___EMPAdaptedTicksLeft -= adaptationReduction;
-			}
-			else
-			{
-			    float adaptationReductionRatio = (adaptationReduction - ___EMPAdaptedTicksLeft) / adaptationReduction;
-			    newStunAdaptedTicks = Mathf.RoundToInt(newStunAdaptedTicks * adaptationReductionRatio);
-			    newStunTicks = Mathf.RoundToInt(newStunTicks * adaptationReductionRatio);
-			    reStun();
-			}
-		    }
-		    		    
-		}
-		else
-		{
-		    __instance.StunFor_NewTmp(Mathf.RoundToInt(dinfo.Amount * 30f), dinfo.Instigator, true, true);
-		    ___EMPAdaptedTicksLeft = Mathf.RoundToInt(dinfo.Amount * 45 * bodySize);
-		    ___stunFromEMP = true;
-		    
-		}
-	    }
-	    return true;
-	}
+        public static bool Prefix(StunHandler __instance, DamageInfo dinfo, bool affectedByEMP, int ___EMPAdaptedTicksLeft, int ___stunTicksLeft, bool ___stunFromEMP)
+        {
+            float bodysize = BUILDING_DEFAULT_BODYSIZE;
+            if (__instance.parent is Pawn)
+            {
+                Pawn p = (Pawn)__instance.parent;
+                if (p.Downed || p.Dead) { return false; }
+                bodysize = p.BodySize;
+            }
+
+            Type t = typeof(StunHandler);
+            FieldInfo EMPAdaptedTicksLeft = t.GetField("EMPAdaptedTicksLeft", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+            FieldInfo StunTicksLeft = t.GetField("stunTicksLeft", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+            FieldInfo StunFromEMP = t.GetField("stunFromEMP", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+            
+            float multiplier = bodysize / Mathf.Pow(bodysize, 2); // Multiplier based on bodysize, smaller for larger creatures
```

Comment:
this is `x / x^2` which reduces to `1 / x`, so you could get rid of the `Mathf.Pow` call and make it
`float multiplier = 1 / bodysize;`
```suggestion
            float multiplier = 1f / bodysize; // Multiplier based on bodysize, smaller for larger creatures
```